Python/C++ API Parity: TransformerDecoderLayer

VinodSKumar opened the pull request:
Fixes #[37756](https://github.com/pytorch/pytorch/issues/37756)

VinodSKumar committed:
CPP TransformerDecoderLayer API

VinodSKumar committed:
CPP TransformerDecoderLayer API

glaringlee reviewed (changes requested):
I have a draft TransformerDecoderLayer here:
https://github.com/pytorch/pytorch/pull/42633

Sorry for post this out late, I was struggling on how to write test cases and where to put them.
Please take a look at my draft, we do have similar stuff there:))). 

I commented you code as much as possible and listed the reasons.

I put all the transformer tests into a new file test/api/transformer.cpp since module.cpp is getting bigger and bigger, we want to separate them into smaller files. but you can leave your test in module.cpp I can move them to transformer.cpp later, but please add CUDA coverage.

It seems your code did not trigger all of our CI test, let's see whether you code can trigger our CI test after you updated your code

glaringlee commented on torch/csrc/api/include/torch/nn/modules/transformer.h:
```diff
@@ -0,0 +1,109 @@
+
+  /// Pretty prints the `TransformerDecoderLayer` module into the given `stream`.
+  /// TODO: To be implemented along with the python API implementation
+  // void pretty_print(std::ostream& stream) const override;
```
Please implement this method if having time. Basically, just printout all the submodules it has

VinodSKumar commented:
> I have a draft TransformerDecoderLayer here:
> #42633
> 
> Sorry for post this out late, I was struggling on how to write test cases and where to put them.
> Please take a look at my draft, we do have similar stuff there:))).
> 
> I commented you code as much as possible and listed the reasons.
> 
> I put all the transformer tests into a new file test/api/transformer.cpp since module.cpp is getting bigger and bigger, we want to separate them into smaller files. but you can leave your test in module.cpp I can move them to transformer.cpp later, but please add CUDA coverage.
> 
> It seems your code did not trigger all of our CI test, let's see whether you code can trigger our CI test after you updated your code

@glaringlee There was some issue with this PR in git, I have created a new PR  [42717](https://github.com/pytorch/pytorch/pull/42717) addressing all the review comments from this PR sorry for the trouble of raising a new PR, could you please help to continue the review on the new PR.
